### PR TITLE
Ack response implement

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
Now acknowledge messages will not get response, but acknowledge messages with transaction will check the acknowledge whether conflict with other transaction in pendingack. We can commit or abort the transaction after we get the response.

## Modification
- We add requestId field in ack protocol and it type is optional.
- Server handle the ack command, if the command carry the requestId, server will return the response.
- In normal, we don't need to get response. Unless we ack with transaction.

## Compatibility
This PR is for transaction, and don't change any client api, so we don't need to think about the compatibility question.

### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (yes)
The rest endpoints: (no)
The admin cli options: (no)
Anything that affects deployment: (no)

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 1]
```diff
@@ -2343,6 +2361,102 @@ private void removeChunkMessage(String msgUUID, ChunkedMessageCtx chunkedMsgCtx,
...
+        long ledgerId;
+        long entryId;
+        if (messageId instanceof BatchMessageIdImpl) {
+            BatchMessageIdImpl batchMessageId = (BatchMessageIdImpl) messageId;
+            bitSetRecyclable = BitSetRecyclable.create();
+            ledgerId = batchMessageId.getLedgerId();
+            entryId = batchMessageId.getEntryId();
+            if (ackType == AckType.Cumulative) {
+                batchMessageId.ackCumulative();
+                bitSetRecyclable.set(0, batchMessageId.getAcker().getBatchSize());
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2378]: The transaction message is a couple of batch messages, make sure can handle this well, @gaoran10 Please help double-check.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 2]
```diff
@@ -2343,6 +2361,102 @@ private void removeChunkMessage(String msgUUID, ChunkedMessageCtx chunkedMsgCtx,
...
+                bitSetRecyclable.clear(0, batchMessageId.getBatchIndex() + 1);
+            } else {
+                batchMessageId.ackIndividual();
+                bitSetRecyclable.set(0, batchMessageId.getAcker().getBatchSize());
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2382]: Same as above comment.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 3]
```diff
@@ -1037,6 +1049,12 @@ private void closeConsumerTasks() {
             possibleSendToDeadLetterTopicMessages.clear();
         }
 
+        if (!ackRequests.isEmpty()) {
+            ackRequests.forEach((key, value) -> value.callback
+                    .completeExceptionally(new TransactionConflictException("Consumer has closed!")));
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1054]: The TransactionConflictException does not make sense here. It's better to add a MessageAcknowledgeException.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 4]
```diff
@@ -2343,6 +2361,102 @@ private void removeChunkMessage(String msgUUID, ChunkedMessageCtx chunkedMsgCtx,
...
+        cmd.retain();
+        cnx().ctx().writeAndFlush(cmd, cnx().ctx().voidPromise());
+        return callBack;
+    }
+
+    protected void ackResponse(CommandAckResponse ackResponse) {
+        checkArgument(ackResponse.getRequestId() >= 0);
+        OpForAckCallBack callBackOp = ackRequests.remove(ackResponse.getRequestId());
+        if (callBackOp == null || callBackOp.callback.isDone()) {
+            log.error("Ack request has been handled requestId : {}", ackResponse.getRequestId());
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2407]: Use info log level here

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 5]
```diff
@@ -2343,6 +2361,102 @@ private void removeChunkMessage(String msgUUID, ChunkedMessageCtx chunkedMsgCtx,
...
+        } else if (!ackResponse.hasError()) {
+            callBackOp.callback.complete(null);
+            if (log.isDebugEnabled()) {
+                log.debug("MessageId : {} has ack by TxnId : {}", callBackOp.messageId.getLedgerId() + ":"
+                        + callBackOp.messageId.getEntryId(), callBackOp.txnID.toString());
+            }
+            callBackOp.recycle();
+        } else {
+            callBackOp.callback.completeExceptionally(new TransactionConflictException(ackResponse.getMessage()));
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:2416]: We should get the specific exception from the `serverError`, should not be throw TransactionConflictException for all errors.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerAckResponseTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static org.mockito.Mockito.*;
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerAckResponseTest.java:42]: Avoid use import .*

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerAckResponseTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                .subscriptionName("sub")
+                .subscriptionType(SubscriptionType.Shared)
+                .ackTimeout(1, TimeUnit.SECONDS)
+                .subscribe();
+        producer.send(1);
+        producer.send(2);
+        try {
+            consumer.acknowledgeAsync(new MessageIdImpl(1, 1, 1), transaction).get();
+        } catch (InterruptedException | ExecutionException e) {
+            Assert.assertTrue(e.getCause() instanceof TransactionConflictException);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerAckResponseTest.java:86]: Why message id (1,1,1) conflict?

**Review by eolivelli — COMMENTED** [review]
It looks like this is a wire protocol change. Is this breaking compatibility with old servers and old clients?

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerAckResponseTest.java (hunk 7 above)
Comment (congbobo184) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerAckResponseTest.java:86]: ledgerId start with 3.

**Review by eolivelli — CHANGES_REQUESTED** [review]
Please state clearly if there are backward compatibility concerns.

In this case I think we should handle it, supporting old clients is a requisite, because we cannot require clients to be updated if you update only the broker.

cc @sijie @merlimat

**Comment by congbobo184** [comment]
@eolivelli I think added protocol fields are all optional type, so it doesn't need to think about compatibility.

**Comment by eolivelli** [comment]
@congbobo184 is the new client able to work with Pulsar 2.6.1 ? it looks like it will expect an ack from the server now.
Probably is just only a matter of documenting better your change.

Can you please describe more this change in the description of the PR ?
- What is the problem that the PR is solving
- What is the design of the solution
- Compatibility with older clients and servers

**Comment by congbobo184** [comment]
@eolivelli This PR is for transaction ack and don't change any client API. In normal ack, we don't change any logical. I change the description of the PR.

**Comment by eolivelli** [comment]
Thank you @congbobo184 
Now it is clear to me.

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 8]
```diff
@@ -201,6 +201,8 @@ enum ServerError {
     TransactionCoordinatorNotFound = 20; // Transaction coordinator not found error
     InvalidTxnStatus = 21; // Invalid txn status error
     NotAllowedError = 22; // Not allowed error
+
+    TransactionConflictException = 23; // Ack with transaction conflict
```
Comment (codelipenghui) [inline pulsar-common/src/main/proto/PulsarApi.proto:205]:
```suggestion
    TransactionConflict = 23; // Transaction conflict
```

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**